### PR TITLE
[Feature/agent_framework] Deprovision API

### LINKED ISSUE #32: [PROPOSAL] Uninstall / Rollback / Overwrite capability for resources
### What are you proposing?

During execution of a workflow, this framework will create resources on the OpenSearch cluster.   Users need the ability to quickly "clean up" these resources for multiple scenarios:
 - After experimenting/prototyping with different approaches or models
 - On partial failure of a workflow

Such resources can include (this is a partial list):
 - Indices
 - ML Models
 - Search and ingest pipelines
 - Workflow templates and sub-templates

We need a capability to:
1. Track what resources have been created in a workflow execution
2. Automate the removal of these resources (if possible)
3. Provide users with a specific list of resources created to enable manual removal in case automated removal fails, or they want more control over the process.

### What problems are you trying to solve?

 - When a user no longer has need for the resources supporting a workflow, we should free up the storage resources allocated to them.
 - When a user iterates on a workflow design, we want to selectively replace only the changed portions.
 - When a user experiences an exception executing a workflow, we want them to have the ability to roll back the cluster to its state prior to the failed execution.

### What is the developer experience going to be?

 - When executing a workflow, the developer should have options to automatically roll back on failure.
 - When no longer using a workflow, the developer should have an API to "uninstall" it completely.
 - The developer should have access to a complete list of resources created to support a workflow so they can roll back manually.

#### Are there any security considerations? 

 - Workflows will require user identity to access some data associated with their execution.  While automated rollback on failure would still have the user identity, a later "uninstall" may not have permission to remove all resources.

### What is the user experience going to be?

**Seeking community feedback!  How do you want this feature to be used?**

### Why should it be built? Any reason not to?

If this feature is not built, developers will need to expend time and effort identifying resources to remove themselves, and/or pay for more storage to maintain unused data.

### What will it take to execute?

In the near term we should keep track of resources created in order to enable manual removal, regardless of any more automatic behavior.

We should research best practices for installation scripts, as this is not a new problem, just a new problem space.

Actual execution will require a detailed design that will be created based on feedback to this proposal.

### Any remaining open questions?

 - Could such a feature be useful as an OpenSearch core feature?   For example, if we had a global parameter (similar to "pretty" and "verbose" tags on URIs) for a "resource group", we could apply that to every request, but that capability would be useful even outside of an automation framework.
 - Any "automated deletion" functionality carries significant risk of inadvertent data loss.  Can/should we integrate this functionality with snapshots / other backup, to enable recovery of an accidental deletion?

### LINKED ISSUE #89: [META] Implement basic manual Undo/Rollback capability
https://github.com/opensearch-project/opensearch-ai-flow-framework/issues/32 describes an overall vision for a full rollback or overwrite capability, and near the bottom says:
> ### What will it take to execute?
> 
> In the near term we should keep track of resources created in order to enable manual removal, regardless of any more automatic behavior.

Which means steps 1 and 3 of:

> We need a capability to:
> 
> 1. Track what resources have been created in a workflow execution
> 2. Automate the removal of these resources (if possible)
> 3. Provide users with a specific list of resources created to enable manual removal in case automated removal fails, or they want more control over the process.

This issue is intended to do exactly that near term requirement.  Review comments on #32 for some more context.

 - [x] Store all resources created in a state index.  https://github.com/opensearch-project/opensearch-ai-flow-framework/issues/74
 - [x] Update Template execution APIs to ensure that this list of resources created is returned to the user as part of the successful response.
 - [x] Update Template execution APIs to ensure that this list of resources created is returned to the user as part of a failed response.
 - [x] Provide a REST API that allows a user to list the resources they created.  This will probably use the same code path which was used in the response generation in the previous steps.
 - [ ] Provide a REST API that allows a user to delete the resources they created.  Look into tagging the resources in such a way that they can be easily filtered/identified.  Consider integrating this as an optional parameter to the delete template API.

### PR DESCRIPTION
**Pull request by dbwiddis**
### Description

Deprovisions resources created by a workflow.

### Issues Resolved

Fixes #89 
Fixes #32

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dbwiddis** [comment]
Current REST response (since we aren't undeploying model it blocks delete model and delete connector:
```json
{
    "error": "Failed to deprovision some resources: [model_id gHcrWowBNatcpxJRFcCN, connector_id fncrWowBNatcpxJRE8BQ]."
}
```

Current output:
```
[2023-12-11T10:37:23,535][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Deprovisioning steps: root_agent_deprovision, sub_agent_deprovision, gpt-3.5-model_deprovision, openAI_connector_deprovision
[2023-12-11T10:37:23,536][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Starting root_agent_deprovision.
[2023-12-11T10:37:23,604][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Successful root_agent_deprovision for agent_id g3crWowBNatcpxJRLMBU
[2023-12-11T10:37:23,604][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Finished root_agent_deprovision.
[2023-12-11T10:37:23,605][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Starting sub_agent_deprovision.
[2023-12-11T10:37:23,663][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Finished sub_agent_deprovision.
[2023-12-11T10:37:23,663][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Successful sub_agent_deprovision for agent_id gncrWowBNatcpxJRK8Cd
[2023-12-11T10:37:23,682][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Starting gpt-3.5-model_deprovision.
[2023-12-11T10:37:23,707][ERROR][o.o.f.w.DeleteModelStep  ] [88665a43e3fa] Failed to delete model
[2023-12-11T10:37:23,708][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Failed gpt-3.5-model_deprovision for model_id gHcrWowBNatcpxJRFcCN: org.opensearch.flowframework.exception.FlowFrameworkException: Model cannot be deleted in deploying or deployed state. Try undeploy model first then delete
[2023-12-11T10:37:23,708][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Starting openAI_connector_deprovision.
[2023-12-11T10:37:23,719][ERROR][o.o.m.a.c.DeleteConnectorTransportAction] [88665a43e3fa] 1 models are still using this connector, please delete or update the models first!
[2023-12-11T10:37:23,720][ERROR][o.o.f.w.DeleteConnectorStep] [88665a43e3fa] Failed to delete connector
[2023-12-11T10:37:23,720][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Failed openAI_connector_deprovision for connector_id fncrWowBNatcpxJRE8BQ: org.opensearch.flowframework.exception.FlowFrameworkException: 1 models are still using this connector, please delete or update the models first: [gHcrWowBNatcpxJRFcCN]
[2023-12-11T10:37:23,721][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Starting gpt-3.5-model_deprovision.
[2023-12-11T10:37:23,723][ERROR][o.o.f.w.DeleteModelStep  ] [88665a43e3fa] Failed to delete model
[2023-12-11T10:37:23,724][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Failed gpt-3.5-model_deprovision for model_id gHcrWowBNatcpxJRFcCN: org.opensearch.flowframework.exception.FlowFrameworkException: Model cannot be deleted in deploying or deployed state. Try undeploy model first then delete
[2023-12-11T10:37:23,724][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Starting openAI_connector_deprovision.
[2023-12-11T10:37:23,726][ERROR][o.o.m.a.c.DeleteConnectorTransportAction] [88665a43e3fa] 1 models are still using this connector, please delete or update the models first!
[2023-12-11T10:37:23,727][ERROR][o.o.f.w.DeleteConnectorStep] [88665a43e3fa] Failed to delete connector
[2023-12-11T10:37:23,727][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Failed openAI_connector_deprovision for connector_id fncrWowBNatcpxJRE8BQ: org.opensearch.flowframework.exception.FlowFrameworkException: 1 models are still using this connector, please delete or update the models first: [gHcrWowBNatcpxJRFcCN]
```

**Comment by dbwiddis** [comment]
Successful deprovision log:
```
[2023-12-12T09:48:22,139][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Deprovisioning steps: root_agent_deprovision, sub_agent_deprovision, deployed-gpt-3.5-model_deprovision, gpt-3.5-model_deprovision, openAI_connector_deprovision
[2023-12-12T09:48:22,140][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Starting root_agent_deprovision.
[2023-12-12T09:48:22,180][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Successful root_agent_deprovision for agent_id 6UkkX4wBMUfZGiR3qNj7
[2023-12-12T09:48:22,180][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Finished root_agent_deprovision.
[2023-12-12T09:48:22,302][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Starting sub_agent_deprovision.
[2023-12-12T09:48:22,342][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Finished sub_agent_deprovision.
[2023-12-12T09:48:22,342][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Successful sub_agent_deprovision for agent_id 6EkkX4wBMUfZGiR3qNhx
[2023-12-12T09:48:22,447][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Starting deployed-gpt-3.5-model_deprovision.
[2023-12-12T09:48:22,447][INFO ][o.o.m.a.u.TransportUndeployModelsAction] [88665a43e3fa] Undeploying model: dhrubo
[2023-12-12T09:48:22,448][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Finished deployed-gpt-3.5-model_deprovision.
[2023-12-12T09:48:22,448][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Successful deployed-gpt-3.5-model_deprovision for model_id 5kkkX4wBMUfZGiR3ktik
[2023-12-12T09:48:22,552][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Starting gpt-3.5-model_deprovision.
[2023-12-12T09:48:22,595][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Successful gpt-3.5-model_deprovision for model_id 5kkkX4wBMUfZGiR3ktik
[2023-12-12T09:48:22,595][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Finished gpt-3.5-model_deprovision.
[2023-12-12T09:48:22,717][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Starting openAI_connector_deprovision.
[2023-12-12T09:48:22,719][ERROR][o.o.m.a.c.DeleteConnectorTransportAction] [88665a43e3fa] 1 models are still using this connector, please delete or update the models first!
[2023-12-12T09:48:22,719][ERROR][o.o.f.w.DeleteConnectorStep] [88665a43e3fa] Failed to delete connector
[2023-12-12T09:48:22,719][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Failed openAI_connector_deprovision for connector_id 5EkkX4wBMUfZGiR3kNjR: org.opensearch.flowframework.exception.FlowFrameworkException: 1 models are still using this connector, please delete or update the models first: [5kkkX4wBMUfZGiR3ktik]
[2023-12-12T09:48:23,723][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Starting openAI_connector_deprovision.
[2023-12-12T09:48:23,766][INFO ][o.o.m.a.c.DeleteConnectorTransportAction] [88665a43e3fa] Completed Delete Connector Request, connector id:5EkkX4wBMUfZGiR3kNjR deleted
[2023-12-12T09:48:23,766][INFO ][o.o.f.w.ProcessNode      ] [88665a43e3fa] Finished openAI_connector_deprovision.
[2023-12-12T09:48:23,766][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] Successful openAI_connector_deprovision for connector_id 5EkkX4wBMUfZGiR3kNjR
[2023-12-12T09:48:24,961][INFO ][o.o.f.t.DeprovisionWorkflowTransportAction] [88665a43e3fa] updated workflow 3UkjX4wBMUfZGiR319gi state to NOT_STARTED
```
Workflow stat was reset. Provisioned and deprovisioned multiple times. :)

File: src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+                ResourceCreated resource = getResourceFromDeprovisionNode(deprovisionNode, resourceMap);
+                String resourceNameAndId = getResourceNameAndId(resource);
+                CompletableFuture<WorkflowData> deprovisionFuture = deprovisionNode.execute();
+                try {
+                    deprovisionFuture.join();
+                    logger.info("Successful {} for {}", deprovisionNode.id(), resourceNameAndId);
+                    // Remove from list so we don't try again
+                    iter.remove();
+                    // Pause briefly before next step
+                    Thread.sleep(100);
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:229]: Curious, why do we need to pause here?

File: src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+                        pn.previousNodeInputs(),
+                        pn.input(),
+                        pn.predecessors(),
+                        this.threadPool,
+                        pn.nodeTimeout()
+                    );
+                }).collect(Collectors.toList());
+                // Pause briefly before next loop
+                try {
+                    Thread.sleep(1000);
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:255]: Also why pause here as well? Just trying to understand better

**Review by joshpalis — APPROVED** [review]
Overall looks good to me, just a few questions for clarification

File: src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+            listener.onFailure(new FlowFrameworkException(e.getMessage(), ExceptionsHelper.status(e)));
+        }
+    }
+
+    private void getResourcesAndExecute(
+        String workflowId,
+        List<ProcessNode> provisionProcessSequence,
+        ActionListener<WorkflowResponse> listener
+    ) {
+        GetWorkflowRequest getRequest = new GetWorkflowRequest(workflowId, true);
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:158]: With PR #273 merged in, this line will have to change to `GetWorkflowStateRequest(workflowId, true)`

File: src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+        client.execute(GetWorkflowAction.INSTANCE, getRequest, ActionListener.wrap(response -> {
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:159]: Additionally here, the action to execute should now be `GetWorkflowStateAction`

**Review by joshpalis — COMMENTED** [review]
@dbwiddis just a heads up

File: src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java (hunk 1 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:229]: Great question that I've asked myself multiple times.

The sequence I've been testing with is "undeploy model" followed by "delete model" followed by "delete connector".

With no delay, the "undeploy" step returns a successful result but attempts to delete a model immediately following sometimes show it's still deployed.  It probably only needs a few milliseconds delay and 100 is excessive. But also short.

There's a similar issue with "delete" model prior to deleting the connector. This seems to have a longer delay over 500ms but always less than 1s.  I've handled this in the "repeat the whole workflow" logic.

File: src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java (hunk 2 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:255]: This is primarily for the "delete model" step that somehow doesn't propogate enough to allow deleting the connector, without at least 500ms but something less than 1s delay.  

Unlike deploying models where we can get a task ID and check for completion, the deletions return immediately (with success) but we have to rely on normal OpenSearch node communication/replication to share the state around.

Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:255]: To summarize the delays (with local testing)

Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:255]: To summarize (with local testing):
1. With no delay: occasional failure after "undeploy" that prevents "delete".  Any short delay resolves. This is probably a "local node" race condition.
2. With no delay: always failure after "delete model" that prevents "delete connector".  Delay over 500ms required.  Delay of 1s alway works. This is probably a "replicate state to all nodes" race condition.

File: src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java (hunk 1 above)
Comment (amitgalitz) [reply src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:231]: Should we give some variable time here, on big clusters where the model is deployed to 30+nodes, I wonder if we will encounter some issue here

File: src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        List<ProcessNode> deprovisionProcessSequence = provisionProcessSequence.stream()
+            // Only include nodes that created a resource
+            .filter(pn -> resourceMap.containsKey(pn.id()))
+            // Create a new ProcessNode with a deprovision step
+            .map(pn -> {
+                String stepName = pn.workflowStep().getName();
+                String deprovisionStep = getDeprovisionStepByWorkflowStep(stepName);
+                // Unimplemented steps presently return null, so skip
+                if (deprovisionStep == null) {
+                    return null;
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:192]: if we don't have an associated deprovision step, will it just skip this?

Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:192]: That's correct, so for example create index won't delete index (yet).

File: src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java (hunk 1 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:231]: I don't think undeploying takes that much time (the responses all return DELETED) so this is just index replication time, which should always be under 1 second, but I don't want a 1 second delay between steps.

Worst case it doesn't delete and they can call it again.

Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:231]: @amitgalitz let me know if we can merge this as is or if you want me to add settings to control the time delays.

Comment (amitgalitz) [reply src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:231]: I am okay merging as is, if it breaks we will fix it :)

File: src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java (hunk 5 above)
Comment (amitgalitz) [reply src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:192]: we can maybe use this concept for deciding not to deprovisiong some things on purpose? For example if i want to delete and deprovision everything but the index

**Review by amitgalitz — APPROVED** [review]

File: src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java (hunk 5 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java:192]: Seems a blunt instrument for that application.    I'll be thinking about update/fine-grained deprovisioning as more APIs become available  -- but this is the point in the code where that might be skipped.

**Pull request merged**